src/main/java/org/opentripplanner/routing/core/RoutingContext.java

```diff
@@ -399,10 +399,32 @@ public boolean isWheelchairAccessible(Vertex v) {
     }
 
     /**
-     * Tear down this routing context, removing any temporary edges.
+     * Tear down this routing context, removing any temporary edges from
+     * the "permanent" graph objects. This enables all temporary objects
+     * for garbage collection.
      */
     public void destroy() {
-        if (origin instanceof TemporaryVertex) ((TemporaryVertex) origin).dispose();
-        if (target instanceof TemporaryVertex) ((TemporaryVertex) target).dispose();
+        disposeTemporaryStart(fromVertex, null);
+        disposeTemporaryEnd(toVertex, null);
+    }
+
+    private static void disposeTemporaryStart(Vertex v, Edge incoming) {
+        if (v instanceof TemporaryVertex) {
+            for (Edge edge : v.getOutgoing()) {
+                disposeTemporaryStart(edge.getToVertex(), edge);
+            }
+        } else if (incoming != null) {
+            v.removeIncoming(incoming);
+        }
+    }
+
+    private static void disposeTemporaryEnd(Vertex v, Edge outgoing) {
+        if (v instanceof TemporaryVertex) {
+            for (Edge edge : v.getIncoming()) {
+                disposeTemporaryEnd(edge.getFromVertex(), edge);
+            }
+        } else if (outgoing != null) {
```

Comment:
I'm trying to wrap my head around the two mutually exclusive cases, where either v is a temporary vertex, or the supplied edge is not null. Is it accurate to say this does a depth-first traversal of the connected subgraph of temporary vertices, and cuts that subgraph off from the main graph at each point it encounters a non-temporary vertex? OTP then holds no references to the temporary subgraph and it is garbage collected.

If this is the mechanism of action we should add Javadoc to these methods explaining it.

Reply:
Yes, that is a very good explanation, I will add it to the JavaDoc.